File: src/main/java/no/nav/pto/veilarbportefolje/oppfolging/VeilederTilordnetService.java
package no.nav.pto.veilarbportefolje.oppfolging;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import no.nav.common.types.identer.AktorId;
import no.nav.common.types.identer.Fnr;
import no.nav.pto.veilarbportefolje.domene.value.VeilederId;
import no.nav.pto.veilarbportefolje.fargekategori.FargekategoriService;
import no.nav.pto.veilarbportefolje.huskelapp.HuskelappService;
import no.nav.pto.veilarbportefolje.kafka.KafkaCommonNonKeyedConsumerService;
import no.nav.pto.veilarbportefolje.opensearch.OpensearchIndexerV2;
import no.nav.pto.veilarbportefolje.persononinfo.PdlIdentRepository;
import org.springframework.stereotype.Service;

import java.time.ZonedDateTime;
import java.util.Optional;

import static no.nav.pto.veilarbportefolje.util.SecureLog.secureLog;


@Slf4j
@Service
@RequiredArgsConstructor
public class VeilederTilordnetService extends KafkaCommonNonKeyedConsumerService<VeilederTilordnetDTO> {
    private final OppfolgingService oppfolgingService;
    private final OppfolgingRepositoryV2 oppfolgingRepositoryV2;
    private final HuskelappService huskelappService;
    private final FargekategoriService fargekategoriService;
    private final OpensearchIndexerV2 opensearchIndexerV2;
    private final PdlIdentRepository pdlIdentRepository;

    @Override
    public void behandleKafkaMeldingLogikk(VeilederTilordnetDTO dto) {
        final AktorId aktoerId = dto.getAktorId();
        final VeilederId veilederId = dto.getVeilederId();
        final ZonedDateTime tilordnetDato = dto.getTilordnet();

        tilordneVeileder(aktoerId, veilederId, tilordnetDato);
    }

    public void tilordneVeileder(AktorId aktoerId, VeilederId veilederId, ZonedDateTime tilordnetDato) {
        oppfolgingRepositoryV2.settVeileder(aktoerId, veilederId);
        oppfolgingRepositoryV2.settTilordningsdato(aktoerId, tilordnetDato);

        kastErrorHvisBrukerSkalVaereUnderOppfolging(aktoerId, veilederId);
        opensearchIndexerV2.oppdaterVeileder(aktoerId, veilederId);
        secureLog.info("Oppdatert bruker: {}, til veileder med id: {}", aktoerId, veilederId);

        Optional<Fnr> maybeFnr = Optional.ofNullable(pdlIdentRepository.hentFnrForAktivBruker(aktoerId));

        final boolean brukerHarByttetNavkontorHuskelapp = huskelappService.brukerHarHuskelappPaForrigeNavkontor(aktoerId, maybeFnr);
        if (brukerHarByttetNavkontorHuskelapp) {
            huskelappService.deaktivereAlleHuskelapperPaaBruker(aktoerId, maybeFnr);
        }

        final boolean brukerHarByttetNavkontorFargekategori = fargekategoriService.brukerHarFargekategoriPaForrigeNavkontor(aktoerId, maybeFnr);
        if (brukerHarByttetNavkontorFargekategori) {
            fargekategoriService.slettFargekategoriPaaBruker(aktoerId, maybeFnr);
        }
    }

    private void kastErrorHvisBrukerSkalVaereUnderOppfolging(AktorId aktorId, VeilederId veilederId) {
        if (hentVeileder(aktorId).equals(veilederId)) {
            return;
        }
        boolean erUnderOppfolgingIVeilarboppfolging = oppfolgingService.hentUnderOppfolging(aktorId);
        if (erUnderOppfolgingIVeilarboppfolging) {
            throw new IllegalStateException("Fikk 'veileder melding' på bruker som enda ikke er under oppfølging i veilarbportefolje");
        }
    }

    private VeilederId hentVeileder(AktorId aktoerId) {
        return oppfolgingRepositoryV2.hentVeilederForBruker(aktoerId)
                .orElse(VeilederId.of(null));
    }
}

File: src/main/java/no/nav/pto/veilarbportefolje/oppfolging/VeilederTilordnetService.kt
package no.nav.pto.veilarbportefolje.oppfolging

import no.nav.common.job.JobRunner
import no.nav.common.types.identer.AktorId
import no.nav.common.types.identer.Fnr
import no.nav.pto.veilarbportefolje.domene.VeilederId
import no.nav.pto.veilarbportefolje.domene.VeilederId.Companion.of
import no.nav.pto.veilarbportefolje.fargekategori.FargekategoriService
import no.nav.pto.veilarbportefolje.huskelapp.HuskelappService
import no.nav.pto.veilarbportefolje.kafka.KafkaCommonNonKeyedConsumerService
import no.nav.pto.veilarbportefolje.opensearch.OpensearchIndexerPaDatafelt
import no.nav.pto.veilarbportefolje.oppfolging.dto.VeilederTilordnetDTO
import no.nav.pto.veilarbportefolje.persononinfo.PdlIdentRepository
import no.nav.pto.veilarbportefolje.util.SecureLog.secureLog
import org.springframework.stereotype.Service
import java.time.ZonedDateTime
import java.util.*

@Service
class VeilederTilordnetService(
    private val oppfolgingClient: OppfolgingClient,
    private val oppfolgingRepositoryV2: OppfolgingRepositoryV2,
    private val huskelappService: HuskelappService,
    private val fargekategoriService: FargekategoriService,
    private val opensearchIndexerPaDatafelt: OpensearchIndexerPaDatafelt,
    private val pdlIdentRepository: PdlIdentRepository
) : KafkaCommonNonKeyedConsumerService<VeilederTilordnetDTO?>() {

    public override fun behandleKafkaMeldingLogikk(dto: VeilederTilordnetDTO?) {
        checkNotNull(dto) { "Ulovlig tilstand: VeilederTilordnetDTO var null. Klarer ikke behandle melding." }

        val aktoerId = dto.aktorId
        val veilederId = dto.veilederId
        val tildeltTidspunkt = dto.tilordnetTidspunkt

        tilordneVeileder(aktoerId, veilederId, tildeltTidspunkt)
    }

    // midlertidig kode for å kjøre batch-jobb
    private val log = org.slf4j.LoggerFactory.getLogger(VeilederTilordnetService::class.java)

    fun lastInnTildelingstidspunktForVeileder() {
        log.info("Startet: Innlastning av tildelingstidspunkt for veileder")

        JobRunner.runAsync("OppfolgingSyncTildelingstidspunkt") {
            val oppfolgingsBrukere =
                oppfolgingRepositoryV2.hentAlleBrukerUnderOppfolgingMedTildeltVeileder()
            oppfolgingsBrukere.forEach(::oppdaterTildelingstidspunkt)

            log.info(
                "OppfolgingsJobb: oppdaterte tildelingstidspunkt pa: {} brukere",
                oppfolgingsBrukere.size
            )
        }

        log.info("Ferdig: Innlastning av tildelingstidspunkt for veileder")
    }

    fun oppdaterTildelingstidspunkt(aktorId: AktorId) {
        try {
            val veilarbInfo = oppfolgingClient.hentVeilarbData(aktorId)

            if (veilarbInfo.isErUnderOppfolging &&
                veilarbInfo.tilordnetTidspunkt != null
            ) {
                oppfolgingRepositoryV2.settTildeltTidspunkt(
                    aktorId,
                    veilarbInfo.tilordnetTidspunkt
                )
            }
        } catch (e: RuntimeException) {
            secureLog.error(
                "RuntimeException i OppfolgingsJobb tildelingstidspunkt for bruker $aktorId",
                e
            )
        } catch (e: Exception) {
            secureLog.error(
                "Exception i OppfolgingsJobb tildelingstidspunkt for bruker $aktorId",
                e
            )
        }
    }

    fun tilordneVeileder(aktoerId: AktorId, veilederId: VeilederId?, tildeltTidspunkt: ZonedDateTime?) {
        oppfolgingRepositoryV2.settVeileder(aktoerId, veilederId)
        oppfolgingRepositoryV2.settTildeltTidspunkt(aktoerId, tildeltTidspunkt)

        kastErrorHvisBrukerSkalVaereUnderOppfolging(aktoerId, veilederId)
        opensearchIndexerPaDatafelt.oppdaterVeileder(aktoerId, veilederId, tildeltTidspunkt)
        secureLog.info("Oppdatert bruker: $aktoerId, til veileder med id: $veilederId, med tildelt tidspunkt: $tildeltTidspunkt")

        val maybeFnr: Fnr? = pdlIdentRepository.hentFnrForAktivBruker(aktoerId)

        val brukerHarByttetNavkontorHuskelapp =
            huskelappService.brukerHarHuskelappPaForrigeNavkontor(aktoerId, Optional.ofNullable(maybeFnr))
        if (brukerHarByttetNavkontorHuskelapp) {
            huskelappService.deaktivereAlleHuskelapperPaaBruker(aktoerId, Optional.ofNullable(maybeFnr))
        }

        val brukerHarByttetNavkontorFargekategori =
            fargekategoriService.brukerHarFargekategoriPaForrigeNavkontor(aktoerId, Optional.ofNullable(maybeFnr))
        if (brukerHarByttetNavkontorFargekategori) {
            fargekategoriService.slettFargekategoriPaaBruker(aktoerId, Optional.ofNullable(maybeFnr))
        }
    }

    private fun kastErrorHvisBrukerSkalVaereUnderOppfolging(aktorId: AktorId?, veilederId: VeilederId?) {
        if (hentVeileder(aktorId) == veilederId) {
            return
        }
        val erUnderOppfolgingIVeilarboppfolging = oppfolgingClient.hentUnderOppfolging(aktorId)
        check(!erUnderOppfolgingIVeilarboppfolging) { "Fikk 'veileder melding' på bruker som enda ikke er under oppfølging i veilarbportefolje" }
    }

    private fun hentVeileder(aktoerId: AktorId?): VeilederId? {
        return oppfolgingRepositoryV2.hentVeilederForBruker(aktoerId)
            .orElse(of(null))
    }
}
